mobile/src/main/java/rs/readahead/washington/mobile/views/fragment/uwazi/widgets/UwaziRelationShipWidget.java

```diff
@@ -0,0 +1,183 @@
+import java.util.Objects;
+
+import rs.readahead.washington.mobile.R;
+import rs.readahead.washington.mobile.presentation.uwazi.UwaziRelationShipEntity;
+import rs.readahead.washington.mobile.views.collect.widgets.QuestionWidget;
+import rs.readahead.washington.mobile.views.custom.CollectRelationShipPreviewView;
+import rs.readahead.washington.mobile.views.fragment.uwazi.entry.UwaziEntryPrompt;
+
+
+@SuppressLint("ViewConstructor")
+public class UwaziRelationShipWidget extends UwaziQuestionWidget {
+    private ArrayList<UwaziRelationShipEntity> relationShipEntities = new ArrayList<>();
+
+    Context context;
+    Button addEntities;
+    ImageButton clearButton;
+    ViewGroup entitiesPanel;
+    PanelToggleButton entitiesToggle;
+    TextView numberOfEntities;
+    View infoEntitiesPanel;
+    private final List<UwaziRelationShipEntity> items;
+
+    public UwaziRelationShipWidget(Context context, @NonNull UwaziEntryPrompt formEntryPrompt, boolean isPdf) {
+        super(context, formEntryPrompt);
+        items = formEntryPrompt.getEntities();
+
+        this.context = context;
+        LinearLayout linearLayout = new LinearLayout(getContext());
+        linearLayout.setOrientation(LinearLayout.VERTICAL);
+
+
+        addImageWidgetViews(linearLayout);
+        addAnswerView(linearLayout);
+        setHelpTextView(getContext().getString(R.string.Uwazi_RelationShipWidget_Select_Entities));
+        clearAnswer();
+        if (items != null) showPreviewFromDraft();
+
+    }
+
+
+    @Override
+    public ArrayList<UwaziRelationShipEntity> getAnswer() {
+        return relationShipEntities;
+
+    }
+
+    @Override
+    public void setFocus(Context context) {
+
+    }
+
+    @Override
+    public String setBinaryData(@NonNull Object data) {
+        ArrayList<UwaziRelationShipEntity> result = new ArrayList<>();
+
+        if (data instanceof String) {
+            ArrayList<UwaziRelationShipEntity> resultList = new Gson().fromJson((String) data, new TypeToken<List<UwaziRelationShipEntity>>() {
+            }.getType());
+            result.addAll(resultList);
+        } else {
+            for (Object o : (ArrayList) data) {
+                if (o instanceof UwaziRelationShipEntity) {
+                    result.add((UwaziRelationShipEntity) o);
+                } else {
+                    String label = Objects.requireNonNull(((LinkedTreeMap) o).get("label")).toString();
+                    String value = Objects.requireNonNull(((LinkedTreeMap) o).get("value")).toString();
+                    result.add(new UwaziRelationShipEntity(value, label, "entity"));
+                }
+            }
+        }
+        if (result != null && !result.isEmpty()) {
+            relationShipEntities = (ArrayList<UwaziRelationShipEntity>) result;
+            if (!relationShipEntities.isEmpty()) {
+                for (UwaziRelationShipEntity relation : relationShipEntities) {
+                    if (!relationShipEntities.contains(relation)) {
```

Comment:
Remove the redundant check for `relationShipEntities.contains(relation)` inside the loop

Reply:
```
     for (Object o : (List<?>) data) {
            if (o instanceof UwaziRelationShipEntity) {
                resultList.add((UwaziRelationShipEntity) o);
            } else {
                String label = Objects.requireNonNull(((Map<String, ?>) o).get("label")).toString();
                String value = Objects.requireNonNull(((Map<String, ?>) o).get("value")).toString();
                resultList.add(new UwaziRelationShipEntity(value, label, "entity"));
            }
        }
```